inc/class-component.php

```diff
@@ -0,0 +1,303 @@
+	 */
+	public function get_examples(): array {
+		return $this->examples;
+	}
+
+	/**
+	 * Gets the name of this component.
+	 *
+	 * @return string The name of this component.
+	 */
+	public function get_name(): string {
+		return $this->name;
+	}
+
+	/**
+	 * Gets the text of the README for this component.
+	 *
+	 * @return string The contents of README.md.
+	 */
+	public function get_readme(): string {
+		return $this->readme;
+	}
+
+	/**
+	 * Gets the array of props for this component.
+	 *
+	 * @return Prop[] An array of Prop objects.
+	 */
+	public function get_props(): array {
+		return $this->props;
+	}
+
+	/**
+	 * Gets the title for this component.
+	 *
+	 * @return string The title for this component.
+	 */
+	public function get_title(): string {
+		return $this->title;
+	}
+
+	/**
+	 * Loads example data from the example at the given index.
+	 *
+	 * @param int $index The index of the example data to load.
+	 */
+	public function load_example_data( int $index ) {
+		// Reset prop values.
+		foreach ( $this->props as $prop ) {
+			$prop->set_value( null );
+		}
+
+		// Load props from examples.
+		if ( isset( $this->examples[ $index ] ) ) {
+			$this->load_props( $this->examples[ $index ]->get_props() );
+		}
+	}
+
+	/**
+	 * Renders this template part, using the props system to validate values
+	 * and provide fallbacks. Mimics the behavior of get_template_part, but
+	 * uses the negotiated path established by locate_component instead of
+	 * WP's built-in template locator. Fires the same hooks as get_template_part
+	 * for increased compatibility with existing hooks and workflows.
+	 */
+	public function render() {
+		// Ensure the template file exists before attempting to render it.
+		if ( ! file_exists( sprintf( '%s/template.php', $this->path ) ) ) {
+			return;
+		}
+
+		// Populate values for all props, using defaults if no value is provided.
+		$props = [];
+		foreach ( $this->props as $prop_name => $prop ) {
+			$props[ $prop_name ] = $prop->get_value();
+		}
+
+		/* phpcs:disable WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound */
+		$slug = sprintf( 'components/%s/template', $this->name );
+		do_action( "get_template_part_{$slug}", $slug, null, $props );
+		do_action( 'get_template_part', $slug, null, [ "{$slug}.php" ], $props );
+		/* phpcs:enable WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound */
+
+		load_template( sprintf( '%s/template.php', $this->path ), false, $props );
+	}
+
+	/**
+	 * Returns an array of directories to look in for the components directory.
+	 * Uses the same logic as the `locate_template` function.
+	 *
+	 * @return array An array of directory paths to look in for the components directory.
+	 */
+	private static function try_dirs(): array {
+		// If we're dogfooding, look in our own component dir.
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( isset( $_GET['dogfooding'] ) && true === (bool) $_GET['dogfooding'] ) {
```

Comment:
Yep, there's a story!

https://en.wikipedia.org/wiki/Eating_your_own_dog_food

Basically, it's the practice of using your own product as a way of testing it. So here, we're using the component manager plugin to test the component manager plugin. By activating this flag, instead of listing the components that are in the _theme_, it lists the components that are in _this plugin_, which are the same components that are used to generate the views in the admin. Basically, rather than just writing a bunch of PHP templates for the admin view, I built the admin view _using components_ and I can build and test those components while writing this plugin in the same way that I would when building and using components in a theme to build a site.

It was a great way to test the behavior and find bugs and things that didn't work well!